File: Day205_2025/Q205_Leetcode.py
import sys
from typing import List


class Solution:
    def minimumScore(self, nums: List[int], edges: List[List[int]]) -> int:
        from collections import defaultdict
        sys.setrecursionlimit(10**6)

        n = len(nums)
        sum_ = [0] * n
        in_ = [0] * n
        out = [0] * n
        time = [0]  # using list for mutable integer
        adj = defaultdict(list)

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        def dfs(x, parent):
            in_[x] = time[0]
            time[0] += 1
            sum_[x] = nums[x]

            for y in adj[x]:
                if y == parent:
                    continue
                dfs(y, x)
                sum_[x] ^= sum_[y]

            out[x] = time[0]

        def calc(a, b, c):
            return max(a, b, c) - min(a, b, c)

        dfs(0, -1)

        res = float('inf')
        for u in range(1, n):
            for v in range(u + 1, n):
                if in_[v] > in_[u] and in_[v] < out[u]:
                    res = min(res, calc(sum_[0] ^ sum_[u], sum_[u] ^ sum_[v], sum_[v]))
                elif in_[u] > in_[v] and in_[u] < out[v]:
                    res = min(res, calc(sum_[0] ^ sum_[v], sum_[v] ^ sum_[u], sum_[u]))
                else:
                    res = min(res, calc(sum_[0] ^ sum_[u] ^ sum_[v], sum_[u], sum_[v]))

        return res

File: Day205_2025/Q205_Leetcode.java
class Q205_Leetcode {
    private int[] sum, in, out;
    private int time;
    private List<List<Integer>> adj;

    private void dfs(int node, int parent, int[] nums) {
        in[node] = time++;
        sum[node] = nums[node];

        for (int neighbor : adj.get(node)) {
            if (neighbor != parent) {
                dfs(neighbor, node, nums);
                sum[node] ^= sum[neighbor];
            }
        }

        out[node] = time;
    }

    private int calc(int a, int b, int c) {
        return Math.max(a, Math.max(b, c)) - Math.min(a, Math.min(b, c));
    }

    public int minimumScore(int[] nums, int[][] edges) {
        int n = nums.length;
        sum = new int[n];
        in = new int[n];
        out = new int[n];
        time = 0;
        adj = new ArrayList<>();

        for (int i = 0; i < n; ++i) {
            adj.add(new ArrayList<>());
        }

        for (int[] e : edges) {
            adj.get(e[0]).add(e[1]);
            adj.get(e[1]).add(e[0]);
        }

        dfs(0, -1, nums);

        int res = Integer.MAX_VALUE;

        for (int u = 1; u < n; ++u) {
            for (int v = u + 1; v < n; ++v) {
                if (in[v] > in[u] && in[v] < out[u]) {
                    res = Math.min(res, calc(sum[0] ^ sum[u], sum[u] ^ sum[v], sum[v]));
                } else if (in[u] > in[v] && in[u] < out[v]) {
                    res = Math.min(res, calc(sum[0] ^ sum[v], sum[v] ^ sum[u], sum[u]));
                } else {
                    res = Math.min(res, calc(sum[0] ^ sum[u] ^ sum[v], sum[u], sum[v]));
                }
            }
        }

        return res;
    }
}
